File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int n = nums.size();
        int l = 1;
        int r = 1;
        int res = nums[0];

        for(int i = 0;i < n;i++){
            l = l == 0 ? 1 : l;
            r = r == 0 ? 1 : r;

            l = l * nums[i];
            r = r * nums[n - 1 - i];

            res = max(res,max(l,r));
        }
        return res;
    }
};

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        l,r,res,n = 1,1,nums[0],len(nums)

        for i in range(n):
            l = 1 if l == 0 else l
            r = 1 if r == 0 else r

            l *= nums[i]
            r *= nums[n - 1 - i]

            res = max(res,max(l,r))
        
        return res


